plutus-core/plutus-core/stdlib/PlutusCore/StdLib/Data/Function.hs

```diff
@@ -36,8 +36,8 @@ import           Control.Monad
 -- | 'id' as a PLC term.
 --
 -- > /\(A :: *) -> \(x : A) -> x
-idFun :: TermLike term TyName Name uni fun => term ()
-idFun = runQuote $ do
+idFun :: TermLike term TyName Name uni fun => Quote (term ())
+idFun = do
```

Comment:
Don't do that. We used to do it like that, but then we realized it was bringing zero safety and only annoyance, so we stopped. There's no difference in terms of safety when it comes to using a *closed* definition. If you have it in `Quote` and unwrap it in a `do` block, nothing prevents you from using the unwrapped definition twice (or using it once and someone else changing it to twice a year later) and then you're screwed, if you don't use `rename` later, and if you do, all that `Quote` dance was pointless. Basically, having closed definitions in `Quote` gives you a false impression of safety and only that. Only open definitions should stay in `Quote`. The global uniqueness invariant should  be satisfied at the call site, we tried to preserve it at the definition site and failed miserably, it's only worth doing that when there are serious performance considerations involve (like in the type checker).

Reply:
Hm. Maybe I need to look at this again. I had to push quite a lot of stuff into `Quote` before I started getting an error in the right place... *something* is not properly closed. Maybe it's just the recursive types stuff.

Reply:
> Previously we had assumed that we would rename such things before inserting them

No, that is way too error-prone. We realized we should not try to preserve global uniqueness modulo exceptional cases. And instead we should satisfy the variant at the call site whenever we need to.

Reply:
> _something_ is not properly closed. 

`fixN` and `fixNAndType` receive an arbitrary `term ()`, so if it's open, the result is also open.

Reply:
Ah, I realised the problem! The paranoid mode in the PIR compiler runs the typechecker a lot, but the _typechecker_ needs global uniqueness. So that's probably why I was getting the wrong error until I incidentally made something that was globally unique already.

Reply:
... and if I rename before typechecking, the error goes away. So perhaps there is no bug, it's just that we needed to rename before typechecking :facepalm:

Reply:
Well, even if there's no bug, I think the change I made introduces less pointless shadowing, so we might as well keep it. I'll get rid of the other stuff, though.

Reply:
> ... and if I rename before typechecking, the error goes away. So perhaps there is no bug, it's just that we needed to rename before typechecking facepalm

Oh, I did not realize what you were saying, but now that I checked the source code, I see that it's in fact perfectly correct, we get that shadowing in the exactly right place. I now take my "sorry" back! Agree with that, though:

> Well, even if there's no bug, I think the change I made introduces less pointless shadowing, so we might as well keep it.